.NET smoke Test Sample

JonathanCrd opened the pull request:
Performs functionalities with Key Vault, Identity, Event Hubs and Blob Storage Track 2 SDKs.

JonathanCrd committed:
.NET smoke Test Sample

Performs functionalities with Key Vault, Identity, Event Hubs and Blob Storage Track 2 SDKs.

danieljurek reviewed (commented):
This is a great first cut! This review covers stylistic and a few things dealing with the general logic of the app without diving too far into program design (except the use of `static`). To that end we'll probably have another review where, after we get the style discussions out of the way, we can dig into the exciting details of Object Oriented design.

JonathanCrd committed:
Env Variables names changed

The names were changed to match the name convention for environment variables.

JonathanCrd committed:
Update .gitignore

JonathanCrd committed:
gitignore updated

A gitignore file was created in the SmokeTest folder, and now all launchSettings.json files are being ignored.

danieljurek commented on samples/SmokeTest/SmokeTest/BlobStorage.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public static async Task performFunctionalities()
+        {
+            Console.WriteLine("\n---------------------------------");
```
Work on this is lower priority given the larger goals of the project. It's worth noting that a unit test framework like xUnit or NUnit can handle this work with little overhead.

danieljurek commented on samples/SmokeTest/SmokeTest/Program.cs:
```diff
@@ -0,0 +1,19 @@
+        {
+            Console.WriteLine("SMOKE TEST FOR TRACK 2 LIBRARIES");
+            await KeyVaultTest.performFunctionalities();
+            await BlobStorage.performFunctionalities();
```
`KeyVaultTest` is a great name. We should standardize the names of these classes so it's clear that their goals are similar. For example `BlobStorage` should be `BlobStorageTest` and `EventHubs` should be `EventHubsTest`

JonathanCrd committed:
Exit Codes, PascalCase and refactoring of the static methods.

Work based on the reviews from the PR #6652. The public methods are now in PascalCase, the methods does not longer return English strings, instead they return booleans or Exceptions.  And Exit Codes handling was implemented.

JonathanCrd committed:
Class names changed

Class names were changed to match the following pattern: <service>Test. Example: BlobStorage -> BlobStorageTest.

JonathanCrd committed:
Comments

XML comments were added and some unnecesary comments were deleted.

JonathanCrd committed:
Higher order functions

Use of higher order functions to dry up the code.

JonathanCrd committed:
Test classes refactured

Drying up the code by using high order functions.

JonathanCrd committed:
Create CosmosDBTest.cs

JonathanCrd committed:
Cosmos DB implementation

Tests done with Mirosoft.Azure.DocumentDB SDK.

JonathanCrd committed:
Update CosmosDBTest.cs

JonathanCrd committed:
Update CosmosDBTest.cs

weshaggard commented on samples/SmokeTest/SmokeTest/BlobStorageTest.cs:
```diff
@@ -0,0 +1,72 @@
+
+namespace SmokeTest
+{
+
```
nit: remove the blank lines.

weshaggard commented on samples/SmokeTest/SmokeTest/TestBase.cs:
```diff
@@ -0,0 +1,32 @@
+        /// </summary>
+        /// <param name="testAction"></param>
+        /// <returns></returns>
+        protected async Task<bool> ExcecuteTest(Func<Task> testAction)
```
typo: `ExcecuteTest` -> `ExecuteTest`

weshaggard commented on samples/SmokeTest/SmokeTest/BlobStorageTest.cs:
```diff
@@ -0,0 +1,72 @@
+            if (!result1)
+            {
+                //If this test failed, then the othe one is going to fail too.
+                Console.WriteLine("Cannot delete the Blob.");
```
I'd suggest you passing in the message to ExecuteTest method and show it in the exception handler in that method.

JonathanCrd replied:
Resolving this review since the high order function is not going to be used anymore.

weshaggard commented:
@JonathanCrd I'd suggest having a look at https://github.com/Azure/azure-sdk-for-net/pull/6678 which is doing something very similar to what you are doing and so the pattern might be interesting here as well.

pakrym commented:
Is the intent here to have a set of samples or to have a set of tests?

weshaggard commented:
>Is the intent here to have a set of samples or to have a set of tests?

The intent is actually to run a set of samples for each of the services in a single application. For now we are just writing samples independently but the goal is to eventually pull in all the samples from the library directory into this single application.

pakrym commented:
The way it's written right now looks like an attempt to reinvent testing framework.

weshaggard commented:
>The way it's written right now looks like an attempt to reinvent testing framework.

I agree and we should definitely avoid duplicating a testing framework. I kind of like what @jsquire did in https://github.com/Azure/azure-sdk-for-net/pull/6678 which is why I pointed to that as an example.

pakrym commented:
I wonder if we can build on that idea but avoid ISample all together. Like, have every file be a self-contained copy-pastable app with Program.Main but also have a common runner that is able to run multiple samples.

jsquire commented:
> I wonder if we can build on that idea but avoid ISample all together. Like, have every file be a self-contained copy-pastable app with Program.Main but also have a common runner that is able to run multiple samples.

I used an interface just to enforce structure;  in the context of Event Hubs, I wanted to ensure that every sample had a consistent method to execute which accepted a common set of connection parameters (to avoid ambient context and the need to set environment variables; I wanted to enable the F5 experience in Visual Studio.)     I could have made implicit assumptions about the structure to avoid the interface, but preferred to make it explicit in case there were community contributions.

With respect to a generic approach that would span client libraries, I'd expect that to be tough without ending up with the need for ambient environment requirements again.  In my opinion, having an approachable "it just works" experience for being able to launch the debugger is the more desirable side of that trade off.

weshaggard commented:
I agree it is something to build on and I'm fine making them standalone but we still need a way to aggregate them together.

pakrym commented:
> With respect to a generic approach that would span client libraries, I'd expect that to be tough without ending up with the need for ambient environment requirements again. In my opinion, having an approachable "it just works" experience for being able to launch the debugger is the more desirable side of that trade off.

I wonder if we can use System.CommandLine (https://github.com/dotnet/command-line-api) inside sample apps themselves making them explicit and directly callable.

JonathanCrd committed:
Unnecessary blank lines removed

JonathanCrd committed:
ExecuteTest typo

JonathanCrd committed:
Base class deleted

The samples not longer need the "TestBase" class to run.

weshaggard commented on samples/SmokeTest/SmokeTest/Program.cs:
```diff
@@ -0,0 +1,33 @@
+        {                        
+            Console.WriteLine("SMOKE TEST FOR TRACK 2 LIBRARIES");
+
+            var keyVault = new KeyVaultTest("SmokeTestSecret", "smokeTestValue", Environment.GetEnvironmentVariable("DIR_TENANT_ID"), Environment.GetEnvironmentVariable("APP_CLIENT_ID"), Environment.GetEnvironmentVariable("CLIENT_SECRET"), Environment.GetEnvironmentVariable("KEY_VAULT_URI"));
```
We should turn these into static method calls and move the configuration into the individual sample file.

JonathanCrd committed:
BlobTestSource.tst deleted

This text file is not going to be needed any more since the test Blob is being created within the code.

JonathanCrd committed:
Static Classes

All classes are now static.

JonathanCrd committed:
Pair programming comments
